### ISSUE TITLE
Add awaitable ThreadPool.Yield()

### ISSUE THREAD

**Issue by davidfowl** [dotnet/corefx#15490]
`Task.Run` breaks the execution flow by forcing the code into a delegate, it's basically callback hell (since it's the first callback in the chain). Instead, for the situations where we want to stay in the await pattern, `ThreadPool.Yield()` would be a nice to have API. It would turn code that looks like this:

```C#
await Task.Run(() => DoSomething());
```

```C#
await ThreadPool.Yield();

// Now on a thread pool thread so we can call some user code
DoSomething();
```

/cc @stephentoub

**Comment by omariom** [comment]
SwitchTo?

**Comment by davidfowl** [comment]
Was keeping in line with `Task.Yield()`

**Comment by stephentoub** [comment]
I agree this is a pretty reasonable thing to add.  In the original async/await CTP, we actually had methods for doing this, IIRC as extensions on a TaskScheduler and SynchronizationContext, e.g.
```C#
await TaskScheduler.Default.SwitchTo(); // continuation will be queued to the thread pool
```
We removed these because you couldn't use an await inside of a catch/finally, which meant it was difficult to return to the original context if you were trying to do something structured like:
```C#
var origScheduler = TaskScheduler.Current;
await TaskScheduler.Default.SwitchTo();
try
{
    ... // runs on thread pool
}
finally { await origScheduler; }
```
and as a result you could end up in situations where error related logic would run on the "wrong" context; in a sense we were concerned we were adding APIs that would lead to dangerous situations.  However, C# has since gained the ability to have awaits in catch/finally blocks, making that largely a moot argument.

I would personally also err on the side of using something more like SwitchTo.  The Yield naming was intended to suggest yielding and coming back to the current context, so using it to mean "go to that other context" might be a little confusing.

**Comment by davidfowl** [comment]
SwitchTo sounds fine to me.

> However, C# has since gained the ability to have awaits in catch/finally blocks, making that largely a moot argument.

If we ever get async disposable maybe that pattern could be more natural.

```C#
await using (TaskScheduler.Current.Preserve())
{
    await TaskScheduler.Default.SwitchTo();
    // Do thread pool stuff
} // Revert to the original scheduler
```

**Comment by benaadams** [comment]
Should it check if already on threadpool/scheduler and immediately complete if so?

**Comment by stephentoub** [comment]
> Should it check if already on threadpool/scheduler and immediately complete if so?

It would depend on your scenario. For example, if you were using it to ensure that a caller of your async API wasn't blocked while doing some long computation, then you'd want it to always queue. But if you were using it to ensure you were running on a context that owned done resources only accessible to that context, you wouldn't care and for perf might want it to mop.

**Comment by jnm2** [comment]
@davidfowl I'd personally prefer this to at least be an option:

```c#
await using (TaskScheduler.Default.SwitchTo())
{
    // Do thread pool stuff
} // Revert to the original scheduler
```

**Comment by noseratio** [comment]
@davidfowl, now have `IAsyncDisposable` and this indeed looks nice:

```
await using (TaskScheduler.Current.Preserve())
{
    await TaskScheduler.Default.SwitchTo();
    // Do thread pool stuff
} // Revert to the original scheduler
```

But I feel it is no different from:

```
await Task.Run(async() => {
    // do thread pool stuff 
    // (or async stuff without sync context)
}); // Revert to the original sync context - without ConfigireAwait(false)
```

Is there any subtle difference?

**Comment by stephentoub** [comment]
> Is there any subtle difference?

Potentially perf, but functionally I'd expect them to be the same.  That said, the former doesn't exist, so it's hard to say for sure.

**Comment by supervos** [comment]
Wouldn't it be possible to have the `SwitchTo` to pass in your `TaskScheduler`. The return value is an async dispobable that contains the previous `TaskScheduler`.

A use case is where you have a limited access to a resource like a database connection or external API service. Together with the async disposable, you can create something like the following:

```csharp
private TaskScheduler _limitedResource = new LimitedTaskScheduler(10);

await using (Task.SwitchTo(_limitedResource))
{
  // Use resource with limited concurrent access
}
```

Placing the `SwitchTo` on the `Task` would make it more visible for people who are not aware of the existence of the `TaskScheduler`.

Another use-case could be to run your code after a `TaskCompletionSource` is completed but you don't know (or can't control) if the continuation runs asynchronously. The `SwitchTo` will prevent the original thread that sets the result of the `TaskCompletionSource` to wait until your code is complete before it can continue with the rest of it's code.

```csharp
await _waiter.Task;
await Task.SwitchTo(TaskScheduler.Default);
// rest of the code
```

**Comment by jnm2** [comment]
Could the API take a cancellation token? Making the SwitchTo implementation aware of cancellation could avoid queuing work on the thread pool when already canceled.

It would be a single call, replacing:

```cs
cancellationToken.ThrowIfCancellationRequested();
await TaskScheduler.Default.SwitchTo();
cancellationToken.ThrowIfCancellationRequested();
```

Usage:

```cs
public static async IAsyncEnumerable<string> EnumerateFilesAsync(
    string path,
    string searchPattern,
    SearchOption searchOption,
    [EnumeratorCancellation] CancellationToken cancellationToken)
{
    await TaskScheduler.Default.SwitchTo(cancellationToken);

    foreach (var file in Directory.EnumerateFiles(path, searchPattern, searchOption))
    {
        yield return file;

        await TaskScheduler.Default.SwitchTo(cancellationToken);
    }
}
```

Maybe not relevant to how .NET Core would implement it, but this implementation actually passes the cancellation token to TaskFactory.StartNew when switching to a non-default task scheduler: https://gist.github.com/jnm2/46a642d2c9f2794ece0b095ba3d96270

**Comment by omariom** [comment]
@noseratio
```C#
await using (TaskScheduler.Current.Preserve())
{
    await TaskScheduler.Default.SwitchTo();
    // Do thread pool stuff
} // Revert to the original scheduler
```
It could be simplified to 
```C#
await using (await TaskScheduler.Default.SwitchTo())
{
}
```

**Comment by jnm2** [comment]
I don't like having `await TaskScheduler.Default.SwitchTo()` produce an IAsyncDisposable. It's much rarer that I'll want to revert to the original scheduler. In the 99% case, I'll get IDE warnings that I'm not disposing a disposable object.

**Comment by omariom** [comment]
It must implement it via async disposable pattern, with `DisposeAsync` returning a struct instead of ValueTask.

**Comment by jnm2** [comment]
The IDE will (should!) still warn when a type follows the async disposable pattern and is not disposed, and I wouldn't like that for this API.

Maybe `_ = await TaskScheduler.Default.SwitchTo();` would be okay, but the name `SwitchTo` doesn't make it obvious what is being discarded.
